docs/settings.md
# Settings

Here are all available settings with their defaults, you can override them in your project settings
```python
DRF_STANDARDIZED_ERRORS = {
    # class responsible for handling the exceptions. Can be subclassed to change
    # which exceptions are handled by default, to update which exceptions are
    # reported to error monitoring tools (like Sentry), ...
    "EXCEPTION_HANDLER_CLASS": "drf_standardized_errors.handler.ExceptionHandler",
    # class responsible for generating error response output. Can be subclassed
    # to change the format of the error response.
    "EXCEPTION_FORMATTER_CLASS": "drf_standardized_errors.formatter.ExceptionFormatter",
    # enable the standardized errors when DEBUG=True for unhandled exceptions.
    # By default, this is set to False so you're able to view the traceback in
    # the terminal and get more information about the exception.
    "ENABLE_IN_DEBUG_FOR_UNHANDLED_EXCEPTIONS": False,
    # When a validation error is raised in a nested serializer, the 'attr' key
    # of the error response will look like:
    # {field}{NESTED_FIELD_SEPARATOR}{nested_field}
    # for example: 'shipping_address.zipcode'
    "NESTED_FIELD_SEPARATOR": ".",

    # The below settings are for OpenAPI 3 schema generation

    # ONLY the responses that correspond to these status codes will appear
    # in the API schema.
    "ALLOWED_ERROR_STATUS_CODES": [
        "400",
        "401",
        "403",
        "404",
        "405",
        "406",
        "415",
        "429",
        "500",
    ],

    # A mapping used to override the default serializers used to describe
    # the error response. The key is the status code and the value is a string
    # that represents the path to the serializer class that describes the
    # error response.
    "ERROR_SCHEMAS": None,

    # When there is a validation error in list serializers, the "attr" returned
    # will be sth like "0.email", "1.email", "2.email", ... So, this setting is
    # used to represent the error codes linked to the same field during API
    # schema generation and its value will be part of the name of the
    # corresponding error component.
    "LIST_INDEX_IN_API_SCHEMA": "INDEX",

    # When there is a validation error in a DictField with the name "extra_data",
    # the "attr" returned will be sth like "extra_data.<key1>", "extra_data.<key2>",
    # "extra_data.<key3>", ... Since the keys of a DictField are not predetermined,
    # this setting is used to represent the error codes linked to the same field
    # during API schema generation and its value will be part of the name of the
    # corresponding error component.
    "DICT_KEY_IN_API_SCHEMA": "KEY",

    # should be unique to error components since it is used to identify error
    # components generated dynamically to exclude them from being processed by
    # the postprocessing hook. This avoids raising warnings for "code" and "attr"
    # which can have the same choices across multiple serializers.
    "ERROR_COMPONENT_NAME_SUFFIX": "ErrorComponent",
}
```


docs/changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [UNRELEASED]
### Fixed
- use a pattern to correctly describe the `attr` value in validation errors in all cases. Specifically, this fixes
the incorrect description of the `attr` value for list serializers and list/dict fields. Previously, the `attr`
value was described with an enum having a single value like `INDEX.field`. Now, it shows up as a string with
the pattern `\d+\.field`.

## [0.16.0] - 2026-04-29
### Added
- Add support for python 3.14
- Add support for django 6.0
- Add support for DRF 3.17

### Changed (backward-incompatible)
- Set the minimum version of drf-spectacular to 0.29.0
- Account for the `format` query parameter raising 404 errors when generating the API schema for error responses. This
  will result in any project using the default value for `REST_FRAMEWORK["URL_FORMAT_OVERRIDE"]` showing 404 errors for
  every operation. That's because DRF enables the `format` query parameter in every endpoint by default. If you don't
  need the `format` query parameter for content negotiation, you can disable it by setting `"URL_FORMAT_OVERRIDE"`
  to `None` in DRF settings. Refer
  to [DRF docs](https://www.django-rest-framework.org/api-guide/settings/#url_format_override) for more information.

## [0.15.0] - 2025-06-09
### Added
- add support for python 3.13
- add support for django 5.2
- add support for DRF 3.16

### Changed (backward-incompatible)
- Unhandled exceptions now return a generic error message by default. This avoids unintentionally leaking
sensitive data included in the exception message. To revert to the old behavior or change the default error
message:
  - create a custom exception handler class
    ```python
    from rest_framework.exceptions import APIException
    from drf_standardized_errors.handler import ExceptionHandler
    
    class MyExceptionHandler(ExceptionHandler):
        def convert_unhandled_exceptions(self, exc: Exception) -> APIException:
            if not isinstance(exc, APIException):
                # `return APIException(detail=str(exc))` restores the old behavior 
                return APIException(detail="New error message")
            else:
                return exc
    ```
  - Then, update the settings to point to your exception handler class
    ```python
    DRF_STANDARDIZED_ERRORS = {
        # ...
        "EXCEPTION_HANDLER_CLASS": "path.to.MyExceptionHandler"
    }
    ```
- set minimum version of drf-spectacular to 0.27.1
- `drf_standardized_errors.types.ErrorType` is now the following type hint
    ```python
    from typing import Literal
    ErrorType = Literal["validation_error", "client_error", "server_error"]
    ```
    `ErrorType` was previously an enum. If you referenced its members in your code, make sure to replace their
    use cases with the newly added constants:
    ```
    from drf_standardized_errors.types import VALIDATION_ERROR, CLIENT_ERROR, SERVER_ERROR
    ErrorType.VALIDATION_ERROR --> VALIDATION_ERROR
    ErrorType.CLIENT_ERROR --> CLIENT_ERROR
    ErrorType.SERVER_ERROR --> SERVER_ERROR
    ```

## [0.14.1] - 2024-08-10
### Added
- declare support for django 5.1

### Fixed
- stop ignoring exceptions with detail as an empty string when returning api errors.

## [0.14.0] - 2024-06-19
### Added
- declare support for DRF 3.15

### Fixed
- enforce support of only drf-spectacular 0.27 and newer in pyproject.toml
- ensure examples from `@extend_schema_serializer` are not ignored when adding error response examples
- show default error response examples only when the corresponding status code is allowed
- add `"null"` to the error code enum of `non_field_errors` validation errors

## [0.13.0] - 2024-02-28
### Changed
- If you're using drf-spectacular 0.27.0 or newer, update `ENUM_NAME_OVERRIDES` entries to reference `choices`
rather than `values`. The list of overrides specific to this package should become like this:
```python
SPECTACULAR_SETTINGS = {
    # other settings
    "ENUM_NAME_OVERRIDES": {
        "ValidationErrorEnum": "drf_standardized_errors.openapi_serializers.ValidationErrorEnum.choices",
        "ClientErrorEnum": "drf_standardized_errors.openapi_serializers.ClientErrorEnum.choices",
        "ServerErrorEnum": "drf_standardized_errors.openapi_serializers.ServerErrorEnum.choices",
        "ErrorCode401Enum": "drf_standardized_errors.openapi_serializers.ErrorCode401Enum.choices",
        "ErrorCode403Enum": "drf_standardized_errors.openapi_serializers.ErrorCode403Enum.choices",
        "ErrorCode404Enum": "drf_standardized_errors.openapi_serializers.ErrorCode404Enum.choices",
        "ErrorCode405Enum": "drf_standardized_errors.openapi_serializers.ErrorCode405Enum.choices",
        "ErrorCode406Enum": "drf_standardized_errors.openapi_serializers.ErrorCode406Enum.choices",
        "ErrorCode415Enum": "drf_standardized_errors.openapi_serializers.ErrorCode415Enum.choices",
        "ErrorCode429Enum": "drf_standardized_errors.openapi_serializers.ErrorCode429Enum.choices",
        "ErrorCode500Enum": "drf_standardized_errors.openapi_serializers.ErrorCode500Enum.choices",
        # other overrides
    },
}
```

### Added
- add compatibility with drf-spectacular 0.27.x
- add support for django 5.0

### Fixed
- Ensure accurate traceback inclusion in 500 error emails sent to ADMINS by capturing the original exception information using `self.exc`. This fixes the issue where tracebacks were previously showing as None for `django version >= 4.1`.
- Handle error responses with +1000 errors

## [0.12.6] - 2023-10-25
### Added
- declare support for type checking
- add support for django 4.2
- add support for python 3.12

### Fixed
- Avoid calling `AutoSchema.get_request_serializer` when inspecting a get operation for possible error responses.

## [0.12.5] - 2023-01-14
### Added

drf_standardized_errors/settings.py
from typing import Any, Dict, Optional, Set, Tuple

from django.conf import settings
from django.core.signals import setting_changed
from django.dispatch import receiver
from rest_framework.settings import import_from_string, perform_import


class PackageSettings:
    """
    Copy of DRF APISettings class with support for importing settings that
    are dicts with value as a string representing the path to the class
    to be imported.
    """

    setting_name = "DRF_STANDARDIZED_ERRORS"

    def __init__(
        self,
        defaults: Optional[Dict[str, Any]] = None,
        import_strings: Optional[Tuple[str, ...]] = None,
    ):
        self.defaults = defaults or DEFAULTS
        self.import_strings = import_strings or IMPORT_STRINGS
        self._cached_attrs: Set[str] = set()

    @property
    def user_settings(self) -> Dict[str, Any]:
        if not hasattr(self, "_user_settings"):
            self._user_settings = getattr(settings, self.setting_name, {})
        return self._user_settings

    def __getattr__(self, attr: str) -> Any:
        if attr not in self.defaults:
            raise AttributeError("Invalid API setting: '%s'" % attr)

        try:
            # Check if present in user settings
            val = self.user_settings[attr]
        except KeyError:
            # Fall back to defaults
            val = self.defaults[attr]

        # Coerce import strings into classes
        if attr in self.import_strings:
            if isinstance(val, dict):
                val = {
                    status_code: import_from_string(error_schema, attr)
                    for status_code, error_schema in val.items()
                }
            else:
                val = perform_import(val, attr)

        # Cache the result
        self._cached_attrs.add(attr)
        setattr(self, attr, val)
        return val

    def reload(self) -> None:
        for attr in self._cached_attrs:
            delattr(self, attr)
        self._cached_attrs.clear()
        if hasattr(self, "_user_settings"):
            delattr(self, "_user_settings")


DEFAULTS: Dict[str, Any] = {
    "EXCEPTION_HANDLER_CLASS": "drf_standardized_errors.handler.ExceptionHandler",
    "EXCEPTION_FORMATTER_CLASS": "drf_standardized_errors.formatter.ExceptionFormatter",
    "ENABLE_IN_DEBUG_FOR_UNHANDLED_EXCEPTIONS": False,
    "NESTED_FIELD_SEPARATOR": ".",
    "ALLOWED_ERROR_STATUS_CODES": [
        "400",
        "401",
        "403",
        "404",
        "405",
        "406",
        "415",
        "429",
        "500",
    ],
    "ERROR_SCHEMAS": None,
    "LIST_INDEX_IN_API_SCHEMA": "INDEX",
    "DICT_KEY_IN_API_SCHEMA": "KEY",
    "ERROR_COMPONENT_NAME_SUFFIX": "ErrorComponent",
}

IMPORT_STRINGS = (
    "EXCEPTION_FORMATTER_CLASS",
    "EXCEPTION_HANDLER_CLASS",
    "ERROR_SCHEMAS",
)

package_settings = PackageSettings(DEFAULTS, IMPORT_STRINGS)


@receiver(setting_changed)
def reload_package_settings(*args: Any, **kwargs: Any) -> None:
    setting = kwargs["setting"]
    if setting == package_settings.setting_name:
        package_settings.reload()
